src/Features/LanguageServer/Protocol/Handler/Diagnostics/DiagnosticSources/DiagnosticSourceManager.cs

```diff
@@ -0,0 +1,133 @@
+    public DiagnosticSourceManager([ImportMany] IEnumerable<IDiagnosticSourceProvider> sourceProviders)
+    {
+        _documentProviders = sourceProviders
+                .Where(p => p.IsDocument)
+                .ToImmutableDictionary(kvp => kvp.Name, kvp => kvp);
+
+        _workspaceProviders = sourceProviders
+                .Where(p => !p.IsDocument)
+                .ToImmutableDictionary(kvp => kvp.Name, kvp => kvp);
+    }
+
+    /// <inheritdoc />
+    public IEnumerable<string> GetSourceNames(bool isDocument)
+        => (isDocument ? _documentProviders : _workspaceProviders).Keys;
+
+    /// <inheritdoc />
+    public async ValueTask<ImmutableArray<IDiagnosticSource>> CreateDiagnosticSourcesAsync(RequestContext context, string? sourceName, bool isDocument, CancellationToken cancellationToken)
+    {
+        var providersDictionary = isDocument ? _documentProviders : _workspaceProviders;
+        if (sourceName != null)
+        {
+            // VS does not distinguish between document and workspace sources. Thus it can request
+            // document diagnostics with workspace source name. We need to handle this case.
+            if (providersDictionary.TryGetValue(sourceName, out var provider))
+                return await provider.CreateDiagnosticSourcesAsync(context, cancellationToken).ConfigureAwait(false);
+
+            context.TraceInformation($"No `{sourceName}` diagnostics for {isDocument}");
+            return [];
+        }
+        else
+        {
+            // VS Code (and legacy VS ?) pass null sourceName when requesting all sources.
+            using var _ = ArrayBuilder<IDiagnosticSource>.GetInstance(out var sourcesBuilder);
+            foreach (var kvp in providersDictionary)
+            {
+                // Exclude Task diagnostics from the aggregated sources.
+                if (kvp.Key != PullDiagnosticCategories.Task)
+                {
+                    var namedSources = await kvp.Value.CreateDiagnosticSourcesAsync(context, cancellationToken).ConfigureAwait(false);
+                    sourcesBuilder.AddRange(namedSources);
+                }
+            }
+
+            var sources = sourcesBuilder.ToImmutableAndClear();
+            return AggregateSourcesIfNeeded(sources, isDocument);
+        }
+    }
+
+    public static ImmutableArray<IDiagnosticSource> AggregateSourcesIfNeeded(ImmutableArray<IDiagnosticSource> sources, bool isDocument)
+    {
+        if (sources.Length <= 1)
+        {
+            return sources;
+        }
+
+        if (isDocument)
+        {
+            // Group all document sources into a single source.
+            Debug.Assert(sources.All(s => s.IsLiveSource()), "All document sources should be live");
+            sources = [new AggregatedDocumentDiagnosticSource(sources)];
+        }
+        else
+        {
+            // For workspace we need to group sources by source id and IsLiveSource
```

Comment:
why do we need to do this?

Reply:
(this comment repeats what the code does, but it doesn't explain why).

Reply:
Same as above - when theres no explicit bucket being requested, we report all diags at once (for all sources).  Those reports need to be grouped by document/project (as a report per doc/project must have all diags for that doc/project).

Reply:
@dibarbet asked for that. I guess it is driven by https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#workspace_diagnostic